// Copyright (c) 2013, Cloudera, inc.
package kudu.tablet;

import "common/wire_protocol.proto";
import "server/metadata.proto";

// Stores the id of the MemRowSet (for inserts or mutations against MRS)
// or of the (row set, delta ID) pair for mutations against a DiskRowSet.
message MemStoreTargetPB {
  // -1 defaults here are so that, if a caller forgets to check has_mrs_id(),
  // they won't accidentally see real-looking (i.e 0) IDs.

  // Either this field...
  optional int64 mrs_id = 1 [ default = -1];

  // ... or both of the following fields are set.
  optional int64 rs_id = 2 [ default = -1 ];
  optional int64 dms_id = 3 [ default = -1 ];
}

// Stores the result of an Insert or Mutate.
message OperationResultPB {
  // set if this particular operation failed
  optional kudu.AppStatusPB failed_status = 2;

  // The stores that the operation affected.
  // For INSERTs, this will always be just one store.
  // For MUTATE, it may be more than one if the mutation arrived during
  // a compaction.
  repeated MemStoreTargetPB mutated_stores = 3;
}

// The final result of a transaction, including the result of each individual
// operation.
message TxResultPB {
  // all the operations in this transaction
  repeated OperationResultPB ops = 1;
}

// Delta statistics for a flushed deltastore
message DeltaStatsPB {
  // Number of deletes (deletes result in deletion of an entire row)
  required int64 delete_count = 1;
  // Repeated update count for each column or zero if no updates are
  // available for the column
  repeated int64 per_column_update_count = 2;
  // The min Timestamp that was stored in this delta.
  required fixed64 min_timestamp = 3;
  // The max Timestamp that was stored in this delta.
  required fixed64 max_timestamp = 4;
}

message TabletStatusPB {
  required string tablet_id = 1;
  required string table_name = 2;
  required metadata.TabletStatePB state = 3;
  required string last_status = 4;
  required string start_key = 5;
  required string end_key = 6;
  optional int64 estimated_on_disk_size = 7;
}
